### ISSUE TITLE
Clean up external dependencies

### ISSUE THREAD

**Issue by breznak** [htm-community/nupic.cpp#148]
During the CMake build phase, we have to find/build several external dependencies or libraries. 
Currently we are using 3 different ways to install them, it might seem smart to unify this behavior: 

The options we can go include: 
1. prerequisities: 
```
  +: the simplest approach for us (nupic.cpp) as we just cmake_find them, 
  +: no (binary) bloating of the repo
  +: (re)installed only once
  +: works offline
  +: "burden" of building is not on us
   +: persists clean builds
  -: user required interaction (one-time install) 
  -: may be harder for us to detect the correct path (user may install in wrong location)
```
(Boost is "installed" this way, as it was too complicated&large to do otherwise) 

2. external sources checked-in the repo in external/ 
```
  -: bloats repo sourcecode with irrelevant files; (still separated in external/) 
  -: longer build time (+ the time for deps) 
  -: burden of building : problems with build (compiler warnings, compiler version differences, ...) 
 +: cmake for us is not too complicated
 +: works offline
```

3. compressed release package with source code added to repo, then built
```
 as 2/
 +: easier to bump version (just replace pkg)
 +: no additional code-lines directly visible in repo
```

4. cmake Fetch downloads & installs externals from repos
```
  as 3,
  +: building better delegated to ustream (their cmakes) 
  -: requires very modern version of cmake, or additional Cmake modules to do so
  -: each build must be connected online
```

Please edit this post and add your +/-, or new methods to install.

**Comment by breznak** [comment]
@dkeeney @ctrl-z-9000-times please take a look when you have time, this is not urgent. 
This is from my POV, so add yours insights please. 

Based on the current, I'd say this repo is not targeted for "even your grandma can do it", so just leave it to users: 1.

**Comment by dkeeney** [comment]
My pref is 4, except for CMake and Boost.  
- I think we need to have tested it with a specific version. Having them install it makes it difficult to know which version.  For example, I could not figure out what version of gtest we had code for.  So I replaced it with the download.  The version we had did not work with C++17 under Windows.
- I don't want to blot our repository so I don't like 2.  In this case its difficult for us to update versions.
- I have less problem with 3 because it is less blot than 2 but there is still blot.  Boost repro would not even fit.
- Number 4 has the least blot.  We control the version. But it does require internet.
- Note that the Download.cmake I now use does not require a modern version of cmake.  It can handle 3 and 4.

Having said all of that.  I will not insist that it be 4.   What about Eigen?

**Comment by dkeeney** [comment]
Oh, the Download.cmake module I found just converts a 4 into a 1 at configure time.

**Comment by ctrl-z-9000-times** [comment]
I like option 4 the best.  I don't think its unreasonable to expect an internet connection for the first compile, after all that is probably the first thing users will do after the downloading this repo.  And after one compile it should all be setup for good. Make-clean doesn't remove it.

**Comment by breznak** [comment]
> What about Eigen? 

- [ ] Eigen: headers only, has releases on GH -> could be installed via FetchModule? https://github.com/eigenteam/eigen-git-mirror/releases 
- [ ] CSV parser: this is currently, one file external/csv.hpp from https://github.com/ben-strasser/fast-cpp-csv-parser ; does not have releases, can Fetch download just from a commit? 
- [ ] do we have a receipe for Boost install this way too? Or we've found it too complicated; or we just don't care as goal is to rm boost eventually as well? 

That would make all our externals' installation unified. 

> Make-clean doesn't remove it. 

But git clean -xdf does. Sometimes you have to do it, esp in these cases: 
- cmake configuration messed up (developing cmake, sometimes only changes in headers, Debug/Release,..) 
- it used to happen with python modules, maybe pybind will be cleaner & safer. 

So you agree to go with 4., :+1: 
I'll be happy if we can find a way to do 4 and keeping an option to configure/install offline somehow...? 
Ie by: 
- make clean target that safely removes all but the downloaded files needed
- find a HowTo to copy the downloaded files into some location and running cmake would pick them up w/o need of re-downloading ..?

**Comment by dkeeney** [comment]
Ok, I will work on this after the current PR is finished.
I think we can have CMake check some directory (perhaps repository/build/downloads) to see if the download is already there before it goes to the internet.  That would provide off-line downloads.

**Comment by ctrl-z-9000-times** [comment]
Wow, Thank you @dkeeney !  
It now downloads all of its dependencies and compiles without complaint.  This is a big improvement.  I also noticed that the CMakeFiles are cleaner & more organized.

**Comment by breznak** [comment]
Possible progress to removing Boost https://github.com/htm-community/nupic.cpp/issues/47#issuecomment-448564459 which is the only dependency that takes long to download and build. 

> Clean up the way external Include dir's and lib's are passed to the main project. 

If possible, please avoid the "super build", I find the current hierarchical approach much simpler. 

All other points are a big :+1:

**Comment by dkeeney** [comment]
I agree the "super build" had some nasty side-affects...like dependencies did not work.

What I am thinking is to use CMake to run a separate process to do all of the 'external' things. That way it can occur at configuration time and dependencies are not needed.  the down side is that it becomes more difficult to pass back references to the main app as to where to find the libs and headers.  So I need to design a way to do that.

**Comment by breznak** [comment]
>  So I need to design a way to do that. 

What would the benefits be? Judge improvement vs. maintainability of the code...

**Comment by dkeeney** [comment]
The goal is to make it more maintainable...and be an improvement.  Better isolation mostly.  The core library build or the bindings should just assume the dependencies are available.

**Comment by dkeeney** [comment]
Perhaps this is the one I should do next.  I don't think it is a big change.  Just moving some things around in the CMake.

**Comment by ctrl-z-9000-times** [comment]
Implemented with PR  #185
